packaging/npm/src/installer.ts

```diff
@@ -0,0 +1,166 @@
+import * as fs from "graceful-fs";
+import * as os from "os";
+import * as path from "path";
+import * as tar from "tar";
+import * as AdmZip from "adm-zip";
+import axios from "axios";
+
+async function downloadAppArchive(
+  version: string,
+  archiveFilename: string,
+  outputPath: string
+): Promise<string> {
+  const repo = "defang-io/defang";
+  const downloadUrl = `https://github.com/${repo}/releases/download/v${version}/${archiveFilename}`;
+  const downloadTargetFile = path.join(outputPath, archiveFilename);
+
+  return downloadFile(downloadUrl, downloadTargetFile);
+}
+
+async function downloadFile(
+  downloadUrl: string,
+  downloadTargetFile: string
+): Promise<string> {
+  try {
+    const writeFileStream = fs.createWriteStream(downloadTargetFile);
+
+    console.log(`Downloading ${downloadUrl}`);
+    const response = await axios.get(downloadUrl, {
+      responseType: "arraybuffer",
+      headers: {
+        "Content-Type": "application/octet-stream",
+      },
+    });
+
+    if (response?.data === undefined) {
+      throw new Error(
+        `Failed to download ${downloadUrl}. No data in response.`
+      );
+    }
+
+    fs.writeFileSync(downloadTargetFile, response.data);
+    return downloadTargetFile;
+  } catch (error) {
+    console.error(`Failed to download ${downloadUrl}. ${error}`);
+    fs.unlinkSync(downloadTargetFile);
+    return "";
+  }
+}
+
+function extractArchive(archiveFilePath: string, outputPath: string): boolean {
+  let result = false;
+
+  console.log(`Extracting ${archiveFilePath}`);
+  const ext = path.extname(archiveFilePath).toLocaleLowerCase();
```

Comment:
windows and macOS use zip and linux uses tar.gz. Uncompress depending on the extension.